tests/e2e-test/internal/runner/runner.go

```diff
@@ -0,0 +1,237 @@
+package runner
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+// Runner runs an e2e test
+type Runner struct {
+	pkgPath string
+	pkgName string
+	network bool
+}
+
+const (
+	expectedDir          string = ".expected"
+	expectedExitCodeFile string = "exitcode.txt"
+	expectedResultsFile  string = "results.yaml"
+	expectedDiffFile     string = "diff.patch"
+	expectedNetworkFile  string = "network.txt"
+)
+
+func isNetworkEnabled(path string) bool {
+	p := filepath.Join(path, expectedDir, expectedNetworkFile)
+	b, err := ioutil.ReadFile(p)
+	if err != nil {
+		return false
+	}
+	if strings.TrimSpace(string(b)) == "true" {
+		return true
+	}
+	return false
+}
+
+// NewRunner returns a new runner for pkg
+func NewRunner(testCase TestCase) (*Runner, error) {
+	pkg := string(testCase)
+	info, err := os.Stat(pkg)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open path %s: %w", pkg, err)
+	}
+	if !info.IsDir() {
+		return nil, fmt.Errorf("path %s is not a directory", pkg)
+	}
+	return &Runner{
+		pkgPath: pkg,
+		pkgName: filepath.Base(pkg),
+		network: isNetworkEnabled(pkg),
+	}, nil
+}
+
+// Run runs the test.
+func (r *Runner) Run() error {
+	fmt.Printf("Running test against package %s\n", r.pkgName)
+	tmpDir, err := ioutil.TempDir("", "kpt-fn-catalog-e2e-*")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary dir: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	tmpPkgPath := filepath.Join(tmpDir, r.pkgName)
+	// create result dir
+	resultsPath := filepath.Join(tmpDir, "results")
+	err = os.Mkdir(resultsPath, 0755)
+	if err != nil {
+		return fmt.Errorf("failed to create results dir %s: %w", resultsPath, err)
+	}
+
+	// copy package to temp directory
+	err = copyDir(r.pkgPath, tmpPkgPath)
+	if err != nil {
+		return fmt.Errorf("failed to copy package: %w", err)
+	}
+
+	// init and commit package files
+	err = r.preparePackage(tmpPkgPath)
+	if err != nil {
+		return fmt.Errorf("failed to prepare package: %w", err)
+	}
+
+	// run function
+	// TODO: change to pipeline when it's ready
+	kptArgs := []string{"fn", "run", tmpPkgPath, "--results-dir", resultsPath}
+	if r.network {
+		kptArgs = append(kptArgs, "--network")
+	}
+	o, fnErr := runCommand("", "kpt", kptArgs)
+	// run formatter
+	_, err = runCommand("", "kpt", []string{"cfg", "fmt", tmpPkgPath})
```

Comment:
Please add a TODO here that links to the GH issue. It should remind us to drop this line after the issue is fixed.

Reply:
I think this line is ok to be left here. Some function may use its own implementation and doesn't respect the original map order. It's better to explicitly run `kpt cfg fmt` after running function.